mage-ai/mage-ai
# [hw] Add support for `dbt-spark`

# Summary


Add the integration support for the `dbt-spark` model.

# Tests


The following procedure has been used to test this change:

1.  Start the Mage development environment through the following command:
```
$  ./scripts/dev.sh demo_project 
```
2. Click the `Terminal` icon on the right side of the Mage UI, and install the required packages and libraries for `pyspark`:
```
root@488dc9529cf3:/home/src#  apt-get update && \
apt-get install -y --no-install-recommends \
        openjdk-11-jre

root@488dc9529cf3:/home/src#  pip install pyspark
```
3. Create a new pipeline with a name `dbt_spark`, and add a `Scratchpad` to test out the connection with PySpark, with the following code:
```
from pyspark.sql import SparkSession
import os

spark = SparkSession.builder.master(os.getenv('SPARK_MASTER_HOST', 'local')).getOrCreate()

spark.sql("show databases;")
```
It should return results similar to the following when running:
```
[Stage 0:>                                                          
(0 + 1) / 1]

    namespace
0 default
```
<img width="696" alt="Screenshot 2023-05-10 at 12 55 52 AM" src="https://github.com/mage-ai/mage-ai/assets/42320565/b016eb70-535b-4ba4-95b7-4fbae0762b14">


For `dbt-spark` testing, a new dbt pipeline `spark_demo` has been created (following instructions in https://docs.mage.ai/tutorials/setup-dbt), with the following settings for `profiles.yml`:
```
spark_demo:
  target: dev
  outputs:
    dev:
      type: spark
      method: session
      schema: default
      host: local
``` 

A new dbt model `model_1.sql` was added to the pipeline with the following code:
```
{{ config(materialized='table') }}

with source_data as (

    select 1 as id
    union all
    select null as id

)

select *
from source_data
```
Click the `Compile & preview` button to execute this new model, which would generate results similar to the following:
![dbt-spark-testing](https://github.com/mage-ai/mage-ai/assets/42320565/10231a21-ed6b-4c0e-9325-be018e630084)

cc:

@wangxiaoyou1993
